File: mobile/lib/services/quicktag.dart
import 'dart:convert';

String quickPathCode(List<String> path) {
  if (path.isEmpty) return '';
  var hash = 0x811c9dc5;
  for (final unit in path.join('\u001f').codeUnits) {
    hash = ((hash ^ (unit & 255)) * 0x01000193) & 0xffffffff;
    hash = ((hash ^ (unit >> 8)) * 0x01000193) & 0xffffffff;
  }
  return hash.toRadixString(36);
}

List<Map<String, dynamic>> quickCategories(List<Map<String, dynamic>> entries,
    {Object? tree, Object? empty}) {
  final nodes = <String, Map<String, dynamic>>{};
  void add(List<String> path) {
    for (var i = 1; i <= path.length; i++) {
      final prefix = path.take(i).toList();
      nodes.putIfAbsent(jsonEncode(prefix),
          () => {'path': prefix, 'code': quickPathCode(prefix), 'count': 0});
    }
  }

  void walk(Object? tree, List<String> parent) {
    if (tree is! List) return;
    for (final node in tree.whereType<Map>()) {
      if (node['name'] is! String || (node['name'] as String).isEmpty) continue;
      final path = [...parent, node['name'] as String];
      add(path);
      walk(node['children'], path);
    }
  }

  walk(tree, []);
  if (empty is List) {
    for (final path in empty.whereType<List>()) {
      if (path.every((p) => p is String && p.isNotEmpty)) {
        add(List<String>.from(path));
      }
    }
  }
  for (final entry in entries) {
    final path = (entry['path'] is List ? entry['path'] as List : [])
        .whereType<String>()
        .where((s) => s.isNotEmpty)
        .toList();
    for (var i = 1; i <= path.length; i++) {
      final prefix = path.take(i).toList(),
          key = jsonEncode(path.take(i).toList());
      final node = nodes.putIfAbsent(key,
          () => {'path': prefix, 'code': quickPathCode(prefix), 'count': 0});
      node['count'] = (node['count'] as int) + 1;
    }
  }
  return nodes.values.toList();
}

bool quickSafe(Map<String, dynamic> entry) =>
    ![true, 1, '1', 'true'].contains(entry['nsfw']) &&
    !RegExp(r'^(r18|r18g|nsfw|explicit|questionable|sensitive|adult)$',
            caseSensitive: false)
        .hasMatch('${entry['rating'] ?? ''}');
bool quickMatch(Map<String, dynamic> entry, String query) {
  final haystack = [
    entry['title'],
    entry['tags'],
    entry['prompt'],
    entry['note'],
    entry['author'],
    entry['credit'],
    ...quickCharacters(entry).map((c) => '${c['label']} ${c['prompt']}'),
    jsonEncode(entry['path'] ?? [])
  ].join(' ').toLowerCase();
  return RegExp(r'-?"[^"]+"|\S+').allMatches(query).every((m) {
    final token = m[0]!, exclude = token.startsWith('-');
    final term = (exclude ? token.substring(1) : token)
        .replaceAll(RegExp(r'^"|"$'), '')
        .toLowerCase();
    return term.isEmpty ||
        (exclude ? !haystack.contains(term) : haystack.contains(term));
  });
}

List<Map<String, String>> quickCharacters(Object? entry) {
  if (entry is! Map || entry['characterPrompts'] is! List) return [];
  return (entry['characterPrompts'] as List)
      .whereType<Map>()
      .where((c) => c['prompt'] is String && (c['prompt'] as String).isNotEmpty)
      .map((c) => {
            'label': c['label'] is String ? c['label'] as String : '',
            'prompt': c['prompt'] as String
          })
      .toList();
}

Map<String, dynamic>? quickLink(String value) {
  if (!RegExp(r'^https?://', caseSensitive: false).hasMatch(value)) return null;
  final url = Uri.parse(value);
  if (url.scheme != 'https' ||
      url.host != 'novelai.quicktagcloud.com' ||
      url.userInfo.isNotEmpty) {
    throw const FormatException('Invalid QuickTagCloud link');
  }
  return {
    'collectionId':
        url.queryParameters['c'] ?? url.queryParameters['codex'] ?? '',
    'path': url.queryParametersAll['path'] ?? <String>[],
    'code': url.queryParameters['p'] ?? '',
    'entry': url.queryParameters['entry'] ??
        Uri.splitQueryString(url.fragment)['entry'] ??
        '',
    'query': url.queryParameters['q'] ?? ''
  };
}

String quickSourceUrl(String collectionId,
        {String entry = '', List<String> path = const []}) =>
    Uri.https('novelai.quicktagcloud.com', '/', {
      'c': collectionId,
      if (entry.isNotEmpty) 'entry': entry,
      if (path.isNotEmpty) 'p': quickPathCode(path)
    }).toString();

String quickCollectionType(Object? type) =>
    type is String && type.isNotEmpty ? type : 'other';
String quickResolveCollection(List<Map<String, dynamic>> catalog, String id) {
  if (id.isEmpty || catalog.any((c) => c['id'] == id)) return id;
  final matches = catalog
      .where((c) => c['aliases'] is List && (c['aliases'] as List).contains(id))
      .toList();
  if (matches.length != 1) {
    throw const FormatException(
        'QuickTagCloud collection alias is unknown or ambiguous');
  }
  return matches.single['id'] as String;
}

Map<String, dynamic> quickCatalogNavigation(
    List<Map<String, dynamic>> catalog, bool safeOnly) {
  final available =
      catalog.where((c) => !safeOnly || c['nsfw'] != true).toList();
  final types = {
    'codex',
    'string',
    'composition',
    'pack',
    ...catalog.map((c) => quickCollectionType(c['type']))
  };
  int entries(Iterable<Map<String, dynamic>> items) => items.fold(
      0, (sum, c) => sum + ((c['entryCount'] as num?)?.toInt() ?? 0));
  return {
    'collections': available
        .map((c) => {
              'id': c['id'],
              'title': c['title'],
              'count': c['entryCount'] ?? 0,
              'type': quickCollectionType(c['type']),
              'version': c['version'] ?? ''
            })
        .toList(),
    'groups': types
        .map((id) => {
              'id': id,
              'count': catalog
                  .where((c) => quickCollectionType(c['type']) == id)
                  .length,
              'visible': available
                  .where((c) => quickCollectionType(c['type']) == id)
                  .length,
              'entries': entries(
                  catalog.where((c) => quickCollectionType(c['type']) == id))
            })
        .where((g) => (g['count'] as int) > 0)
        .toList(),
    'catalogTotal': catalog.length,
    'catalogEntries': entries(catalog),
    'hiddenCollections': catalog.length - available.length,
  };
}

File: src/quicktag.ts
/** QuickTagCloud public URL/data compatibility. No prompt rewriting. */
export interface QuickCategory { path: string[]; code: string; count: number }
export interface QuickNavigation {
  collections: Array<{ id: string; title: string; count: number; type?: string; version?: string }>;
  categories: QuickCategory[];
  categoryPath: string[];
  failedCollections: string[];
  release: string;
  collectionType?: string;
  groups?: Array<{ id: string; count: number; visible: number; entries: number }>;
  catalogTotal?: number; catalogEntries?: number; hiddenCollections?: number;
  declaredCount?: number; loadedCount?: number;
}
export function quickPathCode(path: string[]): string {
  if (!path.length) return "";
  let hash = 0x811c9dc5;
  // Site path codes hash UTF-16LE code units.
  for (const unit of path.join("\u001f").split("").map((c) => c.charCodeAt(0))) {
    hash = Math.imul(hash ^ (unit & 255), 0x01000193);
    hash = Math.imul(hash ^ (unit >>> 8), 0x01000193);
  }
  return (hash >>> 0).toString(36);
}
export function quickCategories(entries: Array<Record<string, unknown>>, tree: unknown = [], empty: unknown = []): QuickCategory[] {
  const nodes = new Map<string, QuickCategory>();
  const add = (path: string[]) => {
    for (let i = 1; i <= path.length; i++) {
      const prefix = path.slice(0, i), key = JSON.stringify(prefix);
      if (!nodes.has(key)) nodes.set(key, { path: prefix, code: quickPathCode(prefix), count: 0 });
    }
  };
  const walk = (value: unknown, parent: string[] = []) => {
    if (!Array.isArray(value)) return;
    for (const node of value) if (node && typeof node.name === "string" && node.name) {
      const path = [...parent, node.name]; add(path); walk(node.children, path);
    }
  };
  walk(tree);
  if (Array.isArray(empty)) for (const path of empty) if (Array.isArray(path) && path.every(p => typeof p === "string" && p)) add(path);
  for (const entry of entries) {
    const path = Array.isArray(entry.path) ? entry.path.filter((s): s is string => typeof s === "string" && !!s) : [];
    for (let i = 1; i <= path.length; i++) {
      const prefix = path.slice(0, i), key = JSON.stringify(prefix);
      const node = nodes.get(key) ?? { path: prefix, code: quickPathCode(prefix), count: 0 };
      node.count++; nodes.set(key, node);
    }
  }
  return [...nodes.values()];
}
export function quickSafe(entry: Record<string, unknown>): boolean {
  return ![true, 1, "1", "true"].some((value) => value === entry.nsfw) && !/^(r18|r18g|nsfw|explicit|questionable|sensitive|adult)$/i.test(String(entry.rating ?? ""));
}
export function quickMatch(entry: Record<string, unknown>, query: string): boolean {
  const haystack = [entry.title, entry.tags, entry.prompt, entry.note, entry.author, entry.credit, JSON.stringify(entry.path ?? []), ...quickCharacters(entry).map((c) => `${c.label} ${c.prompt}`)].join(" ").toLowerCase();
  const tokens = query.match(/-?"[^"]+"|\S+/g) ?? [];
  return tokens.every((token) => { const exclude = token.startsWith("-"); const term = (exclude ? token.slice(1) : token).replace(/^"|"$/g, "").toLowerCase(); return !term || (exclude ? !haystack.includes(term) : haystack.includes(term)); });
}
/** Keep character blocks separate: flattening them changes multi-character prompts. */
export function quickCharacters(entry: unknown): Array<{label: string; prompt: string}> {
  if (!entry || typeof entry !== "object") return [];
  const blocks = (entry as Record<string, unknown>).characterPrompts;
  if (!Array.isArray(blocks)) return [];
  return blocks.flatMap((block) => block && typeof block.prompt === "string" && block.prompt
    ? [{label: typeof block.label === "string" ? block.label : "", prompt: block.prompt}] : []);
}
export function quickLink(value: string): { collectionId: string; path: string[]; code: string; entry: string; query: string } | undefined {
  if (!/^https?:\/\//i.test(value)) return undefined;
  const url = new URL(value);
  if (url.protocol !== "https:" || url.hostname !== "novelai.quicktagcloud.com" || url.username || url.password) throw new Error("Invalid QuickTagCloud link");
  return { collectionId: url.searchParams.get("c") || url.searchParams.get("codex") || "", path: url.searchParams.getAll("path"), code: url.searchParams.get("p") || "", entry: url.searchParams.get("entry") || new URLSearchParams(url.hash.slice(1)).get("entry") || "", query: url.searchParams.get("q") || "" };
}
export function quickSourceUrl(collectionId: string, entry = "", path: string[] = []): string {
  const url = new URL("https://novelai.quicktagcloud.com/"); url.searchParams.set("c", collectionId);
  if (entry) url.searchParams.set("entry", entry);
  if (path.length) url.searchParams.set("p", quickPathCode(path));
  return url.toString();
}

export function quickCollectionType(value: unknown): string {
  return typeof value === "string" && value ? value : "other";
}
export function quickResolveCollection(catalog: Array<Record<string, unknown>>, id: string): string {
  if (!id || catalog.some(c => c.id === id)) return id;
  const matches = catalog.filter(c => Array.isArray(c.aliases) && c.aliases.includes(id));
  if (matches.length !== 1) throw new Error("QuickTagCloud collection alias is unknown or ambiguous");
  return String(matches[0].id);
}
export function quickCatalogNavigation(catalog: Array<Record<string, unknown>>, safeOnly: boolean) {
  const available = catalog.filter(c => !safeOnly || !c.nsfw);
  const types = [...new Set(["codex", "string", "composition", "pack", ...catalog.map(c => quickCollectionType(c.type))])];
  const count = (items: Array<Record<string, unknown>>) => items.reduce((sum, c) => sum + (Number(c.entryCount) || 0), 0);
  return {
    collections: available.map(c => ({ id: String(c.id), title: String(c.title), count: Number(c.entryCount) || 0, type: quickCollectionType(c.type), version: String(c.version ?? "") })),
    groups: types.map(id => ({ id, count: catalog.filter(c => quickCollectionType(c.type) === id).length, visible: available.filter(c => quickCollectionType(c.type) === id).length, entries: count(catalog.filter(c => quickCollectionType(c.type) === id)) })).filter(g => g.count > 0),
    catalogTotal: catalog.length, catalogEntries: count(catalog), hiddenCollections: catalog.length - available.length,
  };
}
